implementations/rust/ockam/ockam_abac/src/policy.rs

```diff
@@ -82,10 +126,134 @@ impl IncomingAccessControl for PolicyAccessControl {
         AbacAccessControl::new(
             self.identities_attributes.clone(),
             self.authority.clone(),
-            policy,
+            policy_expression,
             self.environment.clone(),
         )
         .is_authorized(msg)
         .await
     }
 }
+
+pub struct Policies {
+    resources_repository: Arc<dyn ResourcePoliciesRepository>,
+    resource_types_repository: Arc<dyn ResourceTypePoliciesRepository>,
+}
+
+impl Policies {
+    pub fn new(
+        resources_repository: Arc<dyn ResourcePoliciesRepository>,
+        resource_types_repository: Arc<dyn ResourceTypePoliciesRepository>,
+    ) -> Self {
+        Self {
+            resources_repository,
+            resource_types_repository,
+        }
+    }
+
+    async fn get_expression_for_resource_type_and_name(
+        &self,
+        resource: &Resource,
+        action: &Action,
+    ) -> Result<Expr> {
+        let mut expression = if let Some(policy) = self
+            .resources_repository
+            .get_policy(&resource.resource_name, action)
+            .await?
+        {
+            policy.into_expression()
+        } else {
+            return Err(Error::new(
+                Origin::Application,
+                Kind::Misuse,
+                format!(
+                    "No policy found for resource: {} and action: {}",
+                    resource.resource_name, action
+                ),
+            ));
+        };
+        if let Some(policy) = self
+            .resource_types_repository
+            .get_policy(&resource.resource_type, action)
+            .await?
+        {
+            expression = and([policy.into_expression(), expression]);
+        }
+        Ok(expression)
+    }
+
+    pub async fn set_policy(
```

Comment:
I would just split these functions into 2: 1 for names and 1 for types. But up to you

Reply:
Done + some other changes derived from this idea.